Handle Multi-threaded EGL Context Access

zicklag opened the pull request:
**Connections**
https://github.com/gfx-rs/wgpu/discussions/1630, https://github.com/bevyengine/bevy/issues/841

**Description**
Implements the synchronization necessary to use the GL backend from multiple threads. Accomplishes this by using a mutex around the GL context with extra wrapping to bind and unbind the EGL context when locking and unlocking.

**Testing**
Tested on Ubunty 20.04 with a fork of the Bevy game engine and the WGPU examples ( not that the examples test the multi-threading ).

## Remaining Issues

There is only one Bevy example I cannot get to run yet and it's the `load_gltf` example. It fails with a shader translation error:

```
Jul 26 20:36:50.949 ERROR naga::back::glsl: Conflicting samplers for _group_3_binding_10    
Jul 26 20:36:50.950  WARN wgpu::backend::direct: Shader translation error for stage FRAGMENT: A image was used with multiple samplers    
Jul 26 20:36:50.950  WARN wgpu::backend::direct: Please report it to https://github.com/gfx-rs/naga    
Jul 26 20:36:50.950 ERROR wgpu::backend::direct: wgpu error: Validation Error

Caused by:
    In Device::create_render_pipeline
    Internal error in FRAGMENT shader: A image was used with multiple samplers
```

Interestingly, I think the shader in question doesn't have a `group(3), binding(10)` anywhere that I know of so I'm going to have to drill down a bit more and find out exactly which shader translation is failing more.

This could potentially be fixed in a separate PR. I think the rest of this PR is rather straight-forward and the fix for the error above is probably mostly unrelated to the primary changes made in this PR.

cwfitzgerald commented on wgpu-hal/src/gles/queue.rs:
```diff
@@ -936,7 +939,7 @@ impl crate::Queue<super::Api> for super::Queue {
         for cmd_buf in command_buffers.iter() {
             if let Some(ref label) = cmd_buf.label {
-                self.shared.context.push_debug_group(
+                self.shared.context.lock().push_debug_group(
```
We should probably be favoring locking once at the top of the function and then releasing at the bottom, I can't imagine context switches are terribly cheap.

zicklag replied:
Yeah, I tried that, but the issue was that `self.process` takes `&mut self` and borrowing the `gl` lock at the top takes a reference to `&self.shared.context` so Rust didn't like having the mutable and immutable references to `self` at the same time. I'm not sure a good way around it other than making `process` not take self and moving everything that it nees into function arguments.

Is that what we want to do or is there a better way?

kvark replied:
since `shared` is an arc, we can do:
```rust
let shared = Arc::clone(&self.shared);
let context = shared.context.lock();
```

cwfitzgerald reviewed (commented):
I mean, I hate it, but seems mostly reasonable. I want to test this on stuff like rpi and android to see if it works.

zicklag commented:
> I mean, I hate it, but seems mostly reasonable.

Yeah, according to the EGL spec we can't access the context in multiple threads at a time so essentially we're stuck. :man_shrugging: 

---

Oh, something I forgot. I don't know how to test it, but according to the EGL spec I think the dummy `pbuffer` that we use as a workaround when surfaceless rendering isn't available won't work in a multi-threaded environment. You're not allowed to have multiple threads that bind the same current `pbuffer` I don't think, and that's what would happen if we didn't have surfaceless support and you tried to access the context from multiple threads.

We may need to try to detect that and maybe add a downlevel flag or an error message or something.

cwfitzgerald commented:
We already do have a multithreaded example, it's `cargo test --example hello-compute test_multithreaded_compute`. Additionally, the tests are multi-threaded unless `--test-threads=1` is passed.

Would having thread-local pbuffers work?

I tested this PR on my intel/haswell machine using the GL tests, and got tons of errors:

```
[2021-07-27T06:36:21Z ERROR wgpu_hal::gles::egl] EGL 'eglMakeCurrent' code 0x3009: Got an EGLSurface but no EGLContext
```

jakobhellermann commented:
The shader translation error is probably due to https://github.com/gfx-rs/naga/issues/1137.

kvark commented on wgpu-hal/src/gles/device.rs:
```diff
@@ -74,11 +74,11 @@ impl CompilationContext<'_> {
 impl super::Device {
     unsafe fn compile_shader(
         &self,
+        gl: &glow::Context,
```
this function doesn't need `self` any more

kvark commented on wgpu-hal/src/gles/device.rs:
```diff
@@ -112,6 +112,7 @@ impl super::Device {
 
     fn create_shader(
         &self,
+        gl: &glow::Context,
```
same here

kvark commented on wgpu-hal/src/gles/egl.rs:
```diff
@@ -225,6 +225,77 @@ enum SrgbFrameBufferKind {
+    egl_display: egl::Display,
+    egl_context: egl::Context,
+    egl_pbuffer: Option<egl::Surface>,
+    mutex: Mutex<()>,
```
why not `Mutex<glow::Context>`?

zicklag replied:
Ah, that allows us to still lock the `glow::Context` even when calling `get_without_lock()` which will help a little more to protect against accidental concurrent access. Cool. :)

kvark commented on wgpu-hal/src/gles/egl.rs:
```diff
@@ -225,6 +225,77 @@ enum SrgbFrameBufferKind {
+}
+
+impl AdapterContext {
+    pub fn get_without_lock<'a>(&'a self) -> &'a glow::Context {
```
these kind of lifetimes can be elided

kvark replied:
as we don't want this to be used widely even within the backend, let's make it `unsafe`

kvark commented on wgpu-hal/src/gles/egl.rs:
```diff
@@ -225,6 +225,77 @@ enum SrgbFrameBufferKind {
+            .mutex
+            // Don't lock forever. If it takes longer than 1 second to get the lock we've got a
+            // deadlock and should panic to show where we got stuck
+            .try_lock_for(std::time::Duration::from_secs(1))
```
let's declare a constant at the top of the module

kvark reviewed (commented):
I like this approach quite a bit.
Previously in gfx, we had a wrapper around GL content that would just make it current without any locks. It was quite painful.
Here, we are locking the context and guaranteeing safe access.

zicklag commented:
> Would having thread-local pbuffers work?

I _think_ so. 

> I tested this PR on my intel/haswell machine using the GL tests, and got tons of errors:

Does that machine support surfaceless EGL? If it doesn't, then it's probably the pbfuffer problem. I'll try out thread local pbuffers and then you could test it to see if that fixes it.

> cargo test --example hello-compute test_multithreaded_compute

That one fails on master with the expected "DeviceLost" that you get when creating a buffer on a different thread. It still fails with this PR, but with a different message. Not sure why:

```
---- tests::test_multithreaded_compute stdout ----
thread 'tests::test_multithreaded_compute' panicked at 'UNEXPECTED TEST PASS: BACKEND', wgpu/examples/hello-compute/../../tests/common/mod.rs:299:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

cwfitzgerald commented:
> Does that machine support surfaceless EGL?

It definitely should considering it's running on mesa. Is there a particular extension I should be looking out for? 

> That one fails on master with the expected "DeviceLost" that you get when creating a buffer on a different thread. It still fails with this PR, but with a different message. Not sure why:

We had a test that used to fail on vulkan but now succeeds but we never marked the test to pass. #1731 fixed it.

zicklag commented:
> Is there a particular extension I should be looking out for?

Yeah, in the info log you should see `EGL context: +surfaceless`:

![image](https://user-images.githubusercontent.com/25393315/127196076-b86aa56b-ff59-47f1-874c-b0623c3f10a1.png)

zicklag commented:
> We had a test that used to fail on vulkan but now succeeds but we never marked the test to pass. #1731 fixed it.

For some reason the test still fails after rebasing.

```
$ WGPU_BACKEND=gl cargo test --example hello-compute test_multithreaded_compute
running 1 test
test tests::test_multithreaded_compute ... FAILED

failures:

---- tests::test_multithreaded_compute stdout ----
thread 'tests::test_multithreaded_compute' panicked at 'UNEXPECTED TEST PASS: BACKEND', wgpu/examples/hello-compute/../../tests/common/mod.rs:299:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    tests::test_multithreaded_compute
```

cwfitzgerald commented:
Oh I misread what was causing it. This is because it now passes on your machine but is still marked to fail on OpenGL. You should got to the test definition `wgpu/example/hello-compute/tests.rs` and un-mark it as failing.

> EGL context: +surfaceless

Yeah this doesn't support that, it only supports EGL 1.4. In fact all my GL devices are only EGL 1.4. That being said, it does support `EGL_KHR_surfaceless_context`, so we might need to use that.

zicklag commented:
> Oh I misread what was causing it. This is because it now passes on your machine but is still marked to fail on OpenGL. You should got to the test definition wgpu/example/hello-compute/tests.rs and un-mark it as failing.

Ah, I get it now. :+1:

> Yeah this doesn't support that, it only supports EGL 1.4. In fact all my GL devices are only EGL 1.4. That being said, it does support EGL_KHR_surfaceless_context, so we might need to use that.

Well if it supports the extension then we shouldn't have a problem. Are you not seeing the `+surfaceless` in the logs? Maybe there's something wrong with our check for the extension?

https://github.com/gfx-rs/wgpu/blob/9ce884c384b3df024a78a458f5e659ff4f718580/wgpu-hal/src/gles/egl.rs#L331-L343

cwfitzgerald commented:
Yeah the check is wrong. that should be && not ||.

zicklag commented:
Oh, duh. :man_facepalming: Well, it looks like we found a way to test the pbuffer workaround. :)

I'll fix that check and push an update, then it should work for you on your EGL 1.4 devices. 

Then I'll also experiment to see if I can get the pbuffer workaround working by adding `|| true` to that check so I can pretend I don't have the required version/extensions on my machine.

zicklag commented:
OK I fixed the surfaceless check and I also fixed the error you were getting earlier with "Got an EGLSurface but no EGLContext". When I made the context not current I was not supposed to bind the pbuffer as the current surface.

Also the pbuffer workaround works totally fine with multi-threading so we won't need thread-local pbuffers or anything. I was misinterpreting when we would need to bind the pbuffer. Since we only bind the pbuffer when we bind the context, everything is fine and it is only ever bound by one thread.

If this works on your hardware @cwfitzgerald, I think this should be ready ( pending re-review ).

kvark commented on wgpu-hal/src/gles/egl.rs:
```diff
@@ -1,8 +1,11 @@
 use std::{ffi::CStr, os::raw, ptr, sync::Arc};
 
+/// The amount of time to wait while trying to obtain a lock to the adapter context
+const CONTEXT_LOCK_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(1);
```
let's just define this in seconds?

kvark replied:
let's be consistent with other `std` includes by including either `std::time` or `std::time::Duration`

kvark commented on wgpu-hal/src/gles/egl.rs:
```diff
@@ -329,7 +416,7 @@ impl Inner {
         // and creating dummy pbuffer surface if not.
         let pbuffer =
-            if version < (1, 5) || !display_extensions.contains("EGL_KHR_surfaceless_context") {
+            if version < (1, 5) && !display_extensions.contains("EGL_KHR_surfaceless_context") {
```
uh oh, I think this would look cleaner by reversing the condition entirely
nice that you are fixing this!

kvark commented on wgpu-hal/src/gles/queue.rs:
```diff
@@ -933,25 +940,22 @@ impl crate::Queue<super::Api> for super::Queue {
         signal_fence: Option<(&mut super::Fence, crate::FenceValue)>,
     ) -> Result<(), crate::DeviceError> {
-        self.reset_state();
+        let shared = self.shared.clone();
```
nit: generally speaking, `Arc::clone(&xx)` is preferable over the regular `xx.clone()`

zicklag commented:
Pushed an update!

kvark reviewed (approved):
just pending on @cwfitzgerald

cwfitzgerald commented:
Alright, so all tests pass as long as I pass in `--test-threads=1`. I think the problem may be stemming from having multiple adapters open at once. I need to test my raspberry pi, but I think that is an issue we can deal with in a follow up PR.

zicklag commented:
> Alright, so all tests pass as long as I pass in --test-threads=1. 

Interesting. All the tests pass for me, even without the `--test-threads=1`. I wonder what the difference is.

> I need to test my raspberry pi, but I think that is an issue we can deal with in a follow up PR.

:+1:

kvark commented:
uh, isn't this a bit worrying? The whole point of the PR is to support more than one thread.
Would you have energy to investigate? My time is getting low now since the move is nigh...

cwfitzgerald commented:
> The whole point of the PR is to support more than one thread.

It does, the multi-threaded example works perfectly fine, it's juggling multiple adapters which is the issue (which is what the test harness is doing without --test-threads=1). So I'm comfortable trying to deal with this in a follow up.

> Would you have energy to investigate?

Yeah ofc!

kvark commented:
ok, sounds reasonable
bors r=cwfitzgerald,kvark

zicklag commented:
Awesome! Thanks guys! I'm _super_ excited to actually have gotten Bevy running on OpenGL. :smiley:

cwfitzgerald commented:
Thank you for all the help getting it working!
